### PR TITLE
[Experiments] Add a form control

### LINKED ISSUE #3074: Add a Form Component
We should consider the addition of a Form component, which could wrap a set of inputs and consolidate their values into one dictionary.

Example:
```
    <Form onSubmit={(values: {[key: string]: any}) => console.log(values)}> // {"name": "John Doe", "eula": true, "dob": "3/13/1994"}
        <FormTextInput inputKey="name" validators={[Validators.required()]}/>
        <FormCheckbox inputKey="eula" />
        <FormDatePicker inputKey="dob" />
        <PrimaryButton/>
    </Form>
```

It can use React context to communicate with all of these inputs and gather their values

### PR DESCRIPTION
**Pull request by abettadapur**
#### Pull request checklist

- [x] Addresses an existing issue: Fixes #3074
- [x] Include a change request file using `$ npm run change`

#### Description of changes
* Add a Form component to Experiments
* Enhance the Async.debounce function to allow for cancel and flush operations

### CODE REVIEW

File: packages/utilities/src/Async.ts [hunk 1]
```diff
@@ -373,24 +392,42 @@ export class Async {
...
       return lastResult;
     };
 
+    let cancel = (): void => {
+      if (timeoutId) {
+        this.clearTimeout(timeoutId);
+      }
+    };
+
+    let flush = (): T => {
```
Comment (abettadapur) [inline packages/utilities/src/Async.ts]: Add tests for these new functions

File: packages/utilities/src/Async.ts [hunk 2]
```diff
@@ -444,3 +481,9 @@ export class Async {
     }
   }
 }
+
+export type ICancelable<T> = {
+  flush: () => T;
```
Comment (abettadapur) [inline packages/utilities/src/Async.ts]: Docs

File: packages/experiments/src/components/Form/validators/Validators.ts [hunk 3]
```diff
@@ -0,0 +1,145 @@
+/* tslint:disable:no-any */
```
Comment (abettadapur) [inline packages/experiments/src/components/Form/validators/Validators.ts]: Tests for these Validators

File: packages/experiments/src/components/Form/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,15 @@
+export * from "./Form";
```
Comment (dzearing) [inline packages/experiments/src/components/Form/index.d.ts]: hmm. Why is this a d.ts? This would produce no javascript...

Comment (abettadapur) [reply packages/experiments/src/components/Form/index.d.ts]: Good catch, thanks. Should be .ts

File: packages/experiments/src/components/Form/Form.tsx [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+    submitValue: PropTypes.func.isRequired
+  };
+
+  /**
+   * All registered inputs the form is aware of
+   */
+  private mountedInputs: GenericFormInput[];
+
+  /** Flag which marks whether or not the form has attempted to have been submitted */
+  private pristine: boolean;
```
Comment (dzearing) [inline packages/experiments/src/components/Form/Form.tsx]: privates are underscored.

Comment (abettadapur) [reply packages/experiments/src/components/Form/Form.tsx]: Fixed

**Comment by abettadapur** [comment]
@dzearing @johannao76 @lorejoh12 Are there any more comments on this PR?

**Comment by echarles** [comment]
@abettadapur  @dzearing The [Form Validation example](https://github.com/OfficeDev/office-ui-fabric-react/blob/master/packages/experiments/src/components/Form/examples/Form.Validation.Example.tsx) works will for me but I miss the validation error message.

For now, I implement a double validation:

+ `Validators...` to benefit from the submit button automatic enable/disable.
+ `textFieldProps.onGetErrorMessage` to benefit from the validation error message.

Is there a way to get rid of that double validation and have the message shown to the end-user?

**Comment by echarles** [comment]
@abettadapur @dzearing Just found how to achieve what I want. `showErrorsWhenPristine` property on `Form` does the job. Found this in the reference, but would be good to add it on the demo page also.

**Comment by vertonghenb** [comment]
Good stuff going on here!
I'm looking to implement the Dropdown component to have a multi select as we speak. 
Next up is the Picker.

**Comment by vertonghenb** [comment]
@abettadapur just wondering, the dont's state that we should not manipulate the values, but wouldn't it be great to set initial values for the inputs for example in a Edit form's scenario? I see that the test 'Initial form value' already covers a part of this. Are we planning to implement it?

**Comment by betrue-final-final** [comment]
Dropdown should already have multi select.

**Comment by betrue-final-final** [comment]
https://developer.microsoft.com/en-us/fabric#/components/dropdown

**Comment by abettadapur** [comment]
@betrue-final-final I think he means have the Form wrapper of the Dropdown control implement its multiselect capability

@vertonghenb You can provide initial values to the form inputs today, via the value attribute on the components. I think my point with the don't point was to not treat these as controlled inputs because the Form component is managing the values of the inputs. Does that make sense?

**Comment by vertonghenb** [comment]
@abettadapur Indeed have the multi select working for the wrapper.
About the value it makes sense, also removed the comment as I saw the functionality later.

**Comment by vamsi981** [comment]
Are you guys going to add validations for remaining components like Peoplepickers, toggle etc..or is there any alternative to work with the validations.

**Pull request merged**